File: cpp/tests/unit/test_cfgi_cd_spec.cpp
// ARM SMMU v3 FINDING-NEW-12: CMD_CFGI_CD and CMD_CFGI_CD_ALL command processing
//
// TDD spec tests for ARM §4.3.3 CMD_CFGI_CD and §4.3.4 CMD_CFGI_CD_ALL.
//
// Spec references:
//   §4.3.3  CMD_CFGI_CD (0x05) — Invalidate cached Context Descriptor for (SID, SSID)
//   §4.3.4  CMD_CFGI_CD_ALL (0x06) — Invalidate all cached CDs for SID
//   §6.3.17 SMMU_GERROR — GERROR_CMDQ_ERR must NOT be set for valid commands
//   §7.3.5  C_BAD_STE — must NOT be generated for valid commands

#include <gtest/gtest.h>
#include "smmu/smmu.h"
#include "smmu/types.h"
#include <vector>
#include <memory>

namespace smmu {
namespace test {

// ── Fixture ────────────────────────────────────────────────────────────────

class SMMUCfgiCdTest : public ::testing::Test {
protected:
    void SetUp() override {
        smmu = std::unique_ptr<SMMU>(new SMMU());
        smmu->enable();

        // Configure a stream with Stage 1 translation in Terminate mode
        StreamConfig config;
        config.translationEnabled = true;
        config.stage1Enabled = true;
        config.stage2Enabled = false;
        config.faultMode = FaultMode::Terminate;
        smmu->configureStream(STREAM_ID, config);
        smmu->enableStream(STREAM_ID);
        smmu->createStreamPASID(STREAM_ID, PASID_1);
        smmu->createStreamPASID(STREAM_ID, PASID_2);

        // Map pages for both PASIDs
        PagePermissions perms(true, true, false);
        smmu->mapPage(STREAM_ID, PASID_1, TEST_IOVA, TEST_PA_1, perms);
        smmu->mapPage(STREAM_ID, PASID_2, TEST_IOVA, TEST_PA_2, perms);

        // Enable caching so invalidation commands have observable effect
        smmu->enableCaching(true);

        // Clear any events generated during setup
        smmu->clearEventQueue();
    }

    void TearDown() override {
        smmu.reset();
    }

    std::unique_ptr<SMMU> smmu;

    static constexpr StreamID STREAM_ID   = 0x42;
    static constexpr PASID    PASID_1     = 1;
    static constexpr PASID    PASID_2     = 2;
    static constexpr IOVA     TEST_IOVA   = 0x1000;
    static constexpr PA       TEST_PA_1   = 0x10000;
    static constexpr PA       TEST_PA_2   = 0x20000;

    // Helper: build a CFGI_CD command for (STREAM_ID, pasid)
    CommandEntry makeCfgiCd(PASID pasid) {
        CommandEntry cmd;
        cmd.type      = CommandType::CFGI_CD;
        cmd.streamID  = STREAM_ID;
        cmd.pasid     = pasid;
        return cmd;
    }

    // Helper: build a CFGI_CD_ALL command for STREAM_ID
    CommandEntry makeCfgiCdAll() {
        CommandEntry cmd;
        cmd.type      = CommandType::CFGI_CD_ALL;
        cmd.streamID  = STREAM_ID;
        return cmd;
    }

    // Helper: check event queue for presence of C_BAD_STE
    bool hasCBadSteEvent() {
        auto events = smmu->getEventQueue();
        for (const auto& ev : events) {
            if (ev.type == EventType::C_BAD_STE) {
                return true;
            }
        }
        return false;
    }
};

// ── §4.1.1: Opcode values ─────────────────────────────────────────────────

/// §4.1.1: CMD_CFGI_CD must have opcode 0x05.
TEST_F(SMMUCfgiCdTest, CfgiCdHasCorrectOpcode) {
    EXPECT_EQ(static_cast<int>(CommandType::CFGI_CD), 0x05)
        << "CMD_CFGI_CD must have opcode 0x05 per ARM IHI0070G.b §4.1.1";
}

/// §4.1.1: CMD_CFGI_CD_ALL must have opcode 0x06.
TEST_F(SMMUCfgiCdTest, CfgiCdAllHasCorrectOpcode) {
    EXPECT_EQ(static_cast<int>(CommandType::CFGI_CD_ALL), 0x06)
        << "CMD_CFGI_CD_ALL must have opcode 0x06 per ARM IHI0070G.b §4.1.1";
}

// ── §6.3.17: GERROR must not be set for valid commands ────────────────────

/// §6.3.17: Processing CMD_CFGI_CD must NOT set GERROR_CMDQ_ERR.
TEST_F(SMMUCfgiCdTest, CfgiCdDoesNotSetGerror) {
    smmu->submitCommand(makeCfgiCd(PASID_1));
    smmu->processCommandQueue();

    EXPECT_EQ(smmu->getGerror() & GERROR_CMDQ_ERR, 0u)
        << "CMD_CFGI_CD is a valid command and must not set GERROR_CMDQ_ERR";
}

/// §6.3.17: Processing CMD_CFGI_CD_ALL must NOT set GERROR_CMDQ_ERR.
TEST_F(SMMUCfgiCdTest, CfgiCdAllDoesNotSetGerror) {
    smmu->submitCommand(makeCfgiCdAll());
    smmu->processCommandQueue();

    EXPECT_EQ(smmu->getGerror() & GERROR_CMDQ_ERR, 0u)
        << "CMD_CFGI_CD_ALL is a valid command and must not set GERROR_CMDQ_ERR";
}

// ── §7.3.5: C_BAD_STE must not be generated for valid commands ───────────

/// §7.3.5: Processing CMD_CFGI_CD must NOT generate a C_BAD_STE event.
TEST_F(SMMUCfgiCdTest, CfgiCdDoesNotGenerateBadSteEvent) {
    smmu->submitCommand(makeCfgiCd(PASID_1));
    smmu->processCommandQueue();

    EXPECT_FALSE(hasCBadSteEvent())
        << "CMD_CFGI_CD must not produce a C_BAD_STE event";
}

/// §7.3.5: Processing CMD_CFGI_CD_ALL must NOT generate a C_BAD_STE event.
TEST_F(SMMUCfgiCdTest, CfgiCdAllDoesNotGenerateBadSteEvent) {
    smmu->submitCommand(makeCfgiCdAll());
    smmu->processCommandQueue();

    EXPECT_FALSE(hasCBadSteEvent())
        << "CMD_CFGI_CD_ALL must not produce a C_BAD_STE event";
}

// ── §4.3.3: CMD_CFGI_CD invalidates the (SID, SSID) CD cache ─────────────

/// §4.3.3: After CMD_CFGI_CD the stream/PASID can still translate correctly.
/// Invalidation must flush the cache entry but not destroy the page mapping.
TEST_F(SMMUCfgiCdTest, CfgiCdInvalidatesPasidCache) {
    // Warm the cache with a successful translation
    TranslationResult before = smmu->translate(STREAM_ID, PASID_1, TEST_IOVA,
                                               AccessType::Read, SecurityState::NonSecure);
    ASSERT_TRUE(before.isOk()) << "Pre-invalidation translation must succeed";

    // Issue CMD_CFGI_CD to flush the cache entry for (STREAM_ID, PASID_1)
    smmu->submitCommand(makeCfgiCd(PASID_1));
    smmu->processCommandQueue();

    // The page mapping is still present — translation must still succeed
    TranslationResult after = smmu->translate(STREAM_ID, PASID_1, TEST_IOVA,
                                              AccessType::Read, SecurityState::NonSecure);
    EXPECT_TRUE(after.isOk())
        << "Translation must still succeed after CMD_CFGI_CD (mapping is intact)";
    EXPECT_EQ(smmu->getGerror() & GERROR_CMDQ_ERR, 0u);
}

/// §4.3.4: After CMD_CFGI_CD_ALL both PASIDs in the stream still translate.
TEST_F(SMMUCfgiCdTest, CfgiCdAllInvalidatesAllPasids) {
    // Warm cache for both PASIDs
    ASSERT_TRUE(smmu->translate(STREAM_ID, PASID_1, TEST_IOVA,
                                AccessType::Read, SecurityState::NonSecure).isOk());
    ASSERT_TRUE(smmu->translate(STREAM_ID, PASID_2, TEST_IOVA,
                                AccessType::Read, SecurityState::NonSecure).isOk());

    // Issue CMD_CFGI_CD_ALL — flushes all CD entries for STREAM_ID
    smmu->submitCommand(makeCfgiCdAll());
    smmu->processCommandQueue();

    // Mappings still exist — both translations must still succeed
    EXPECT_TRUE(smmu->translate(STREAM_ID, PASID_1, TEST_IOVA,
                                AccessType::Read, SecurityState::NonSecure).isOk())
        << "PASID_1 translation must succeed after CMD_CFGI_CD_ALL";
    EXPECT_TRUE(smmu->translate(STREAM_ID, PASID_2, TEST_IOVA,
                                AccessType::Read, SecurityState::NonSecure).isOk())
        << "PASID_2 translation must succeed after CMD_CFGI_CD_ALL";
}

// ── §4.3.3: CMD_CFGI_CD is PASID-scoped ───────────────────────────────────

/// §4.3.3: CMD_CFGI_CD for PASID_1 must not affect PASID_2 translations.
TEST_F(SMMUCfgiCdTest, CfgiCdForWrongPasidDoesNotAffectOtherPasid) {
    // Warm cache for both PASIDs
    ASSERT_TRUE(smmu->translate(STREAM_ID, PASID_1, TEST_IOVA,
                                AccessType::Read, SecurityState::NonSecure).isOk());
    ASSERT_TRUE(smmu->translate(STREAM_ID, PASID_2, TEST_IOVA,
                                AccessType::Read, SecurityState::NonSecure).isOk());

    // Invalidate only PASID_1's CD entry
    smmu->submitCommand(makeCfgiCd(PASID_1));
    smmu->processCommandQueue();

    // PASID_2's mapping is unaffected — must still translate successfully
    EXPECT_TRUE(smmu->translate(STREAM_ID, PASID_2, TEST_IOVA,
                                AccessType::Read, SecurityState::NonSecure).isOk())
        << "CMD_CFGI_CD for PASID_1 must not disrupt PASID_2 translations";
}

// ── §4.3.3 / §4.3.4: Command followed by SYNC ────────────────────────────

/// CMD_CFGI_CD followed by CMD_SYNC must process without errors.
TEST_F(SMMUCfgiCdTest, CfgiCdFollowedBySync) {
    smmu->submitCommand(makeCfgiCd(PASID_1));

    CommandEntry sync;
    sync.type = CommandType::SYNC;
    smmu->submitCommand(sync);

    smmu->processCommandQueue();

    EXPECT_EQ(smmu->getGerror() & GERROR_CMDQ_ERR, 0u)
        << "CMD_CFGI_CD + CMD_SYNC must not set GERROR_CMDQ_ERR";
    EXPECT_FALSE(hasCBadSteEvent())
        << "CMD_CFGI_CD + CMD_SYNC must not produce C_BAD_STE";
}

/// CMD_CFGI_CD_ALL followed by CMD_SYNC must process without errors.
TEST_F(SMMUCfgiCdTest, CfgiCdAllFollowedBySync) {
    smmu->submitCommand(makeCfgiCdAll());

    CommandEntry sync;
    sync.type = CommandType::SYNC;
    smmu->submitCommand(sync);

    smmu->processCommandQueue();

    EXPECT_EQ(smmu->getGerror() & GERROR_CMDQ_ERR, 0u)
        << "CMD_CFGI_CD_ALL + CMD_SYNC must not set GERROR_CMDQ_ERR";
    EXPECT_FALSE(hasCBadSteEvent())
        << "CMD_CFGI_CD_ALL + CMD_SYNC must not produce C_BAD_STE";
}

// ── §4.3.3: Unknown stream is a no-op ────────────────────────────────────

/// §4.3.3: CMD_CFGI_CD for an unconfigured StreamID must be a no-op (no GERROR, no events).
TEST_F(SMMUCfgiCdTest, CfgiCdUnknownStreamIsNoOp) {
    CommandEntry cmd;
    cmd.type     = CommandType::CFGI_CD;
    cmd.streamID = 0xDEAD; // stream never configured
    cmd.pasid    = PASID_1;

    smmu->submitCommand(cmd);
    smmu->processCommandQueue();

    EXPECT_EQ(smmu->getGerror() & GERROR_CMDQ_ERR, 0u)
        << "CMD_CFGI_CD for unknown stream must not set GERROR_CMDQ_ERR";
    EXPECT_FALSE(hasCBadSteEvent())
        << "CMD_CFGI_CD for unknown stream must not generate C_BAD_STE";
}

} // namespace test
} // namespace smmu

File: rust/smmu/tests/test_cfgi_cd_spec.rs
#![allow(clippy::doc_markdown)]
#![allow(clippy::uninlined_format_args)]

//! FINDING-H-03: CMD_CFGI_CD and CMD_CFGI_CD_ALL command types.
//!
//! Spec: ARM IHI0070G.b §4.3.3 (CMD_CFGI_CD, opcode 0x05),
//!       §4.3.4 (CMD_CFGI_CD_ALL, opcode 0x06)
//!
//! Requirements:
//! - CommandType::CfgiCd  must exist with opcode 0x05.
//! - CommandType::CfgiCdAll must exist with opcode 0x06.
//! - CMD_CFGI_CD invalidates all TLB entries cached from the Context Descriptor
//!   of the specified (stream_id, PASID / SubstreamID) pair.
//! - CMD_CFGI_CD_ALL invalidates all TLB entries cached from every Context
//!   Descriptor of the specified stream (all PASIDs).
//! - TLB entries for other streams / PASIDs must survive both commands.

use smmu::types::{
    AccessType, CommandEntry, CommandType, PagePermissions, SecurityState, StreamConfig, StreamID,
    IOVA, PA, PASID,
};
use smmu::SMMU;

// ── Helpers ───────────────────────────────────────────────────────────────────

fn make_smmu() -> SMMU {
    let smmu = SMMU::new();
    smmu.enable().unwrap();
    smmu
}

fn sid(n: u32) -> StreamID {
    StreamID::new(n).unwrap()
}
fn pasid(n: u32) -> PASID {
    PASID::new(n).unwrap()
}
fn iova(addr: u64) -> IOVA {
    IOVA::new(addr).unwrap()
}
fn pa(addr: u64) -> PA {
    PA::new(addr).unwrap()
}

/// Configure a Stage-1 stream, create `num_pasids` PASIDs each with one mapped page,
/// then translate each once so TLB entries are warmed up.
fn setup_stream_with_pasids(smmu: &SMMU, stream_n: u32, num_pasids: u32) {
    let s = sid(stream_n);
    let config = StreamConfig::builder()
        .stage1_enabled(true)
        .translation_enabled(true)
        .security_enforced(false)
        .pasid_enabled(true)
        .max_pasid(num_pasids)
        .build()
        .unwrap();
    smmu.configure_stream(s, config).unwrap();

    for p in 0..num_pasids {
        let ps = pasid(p);
        smmu.create_pasid(s, ps).unwrap();
        let iova_addr = 0x1000 + u64::from(p) * 0x1000;
        let pa_addr = 0x8000_0000 + u64::from(p) * 0x1000;
        smmu.map_page(s, ps, iova(iova_addr), pa(pa_addr), PagePermissions::read_write(), SecurityState::NonSecure).unwrap();
        // Warm up TLB
        let _ = smmu.translate(s, ps, iova(iova_addr), AccessType::Read, SecurityState::NonSecure);
    }
}

/// Returns true when a translation hits the TLB (i.e., the page is still cached/mapped).
fn is_cached(smmu: &SMMU, stream_n: u32, pasid_n: u32) -> bool {
    let iova_addr = 0x1000 + u64::from(pasid_n) * 0x1000;
    smmu.translate(sid(stream_n), pasid(pasid_n), iova(iova_addr), AccessType::Read, SecurityState::NonSecure).is_ok()
}

/// Build a zero-init CommandEntry of the given type, overriding stream_id and pasid.
const fn cfgi_cd_cmd(stream_id: u32, pasid_val: u32) -> CommandEntry {
    CommandEntry {
        cmd_type: CommandType::CfgiCd,
        stream_id,
        pasid: pasid_val,
        start_address: 0,
        end_address: 0,
        flags: 0,
        timestamp: 0,
        asid: 0,
        vmid: 0,
        stag: 0,
        prg_index: 0,
        action: false,
        abort: false,
        range: 31,
        leaf: false,
        cs: 0,
        security_state: SecurityState::NonSecure,
    }
}

const fn cfgi_cd_all_cmd(stream_id: u32) -> CommandEntry {
    CommandEntry {
        cmd_type: CommandType::CfgiCdAll,
        stream_id,
        pasid: 0,
        start_address: 0,
        end_address: 0,
        flags: 0,
        timestamp: 0,
        asid: 0,
        vmid: 0,
        stag: 0,
        prg_index: 0,
        action: false,
        abort: false,
        range: 31,
        leaf: false,
        cs: 0,
        security_state: SecurityState::NonSecure,
    }
}

// ── Opcode correctness ────────────────────────────────────────────────────────

/// CMD_CFGI_CD must have opcode 0x05 per ARM §4.3.3 / §4.1.1.
#[test]
fn test_cfgi_cd_opcode_is_0x05() {
    assert_eq!(CommandType::CfgiCd as u8, 0x05,
               "CMD_CFGI_CD opcode must be 0x05 per ARM IHI0070G.b §4.1.1");
}

/// CMD_CFGI_CD_ALL must have opcode 0x06 per ARM §4.3.4 / §4.1.1.
#[test]
fn test_cfgi_cd_all_opcode_is_0x06() {
    assert_eq!(CommandType::CfgiCdAll as u8, 0x06,
               "CMD_CFGI_CD_ALL opcode must be 0x06 per ARM IHI0070G.b §4.1.1");
}

// ── CMD_CFGI_CD — single PASID invalidation ───────────────────────────────────

/// CMD_CFGI_CD for (stream=1, pasid=0) must invalidate TLB entries for that
/// specific (stream, PASID) pair.
#[test]
fn test_cfgi_cd_invalidates_target_pasid_tlb() {
    let smmu = make_smmu();
    setup_stream_with_pasids(&smmu, 1, 2); // pasid 0 and 1

    // Confirm both are cached / accessible
    assert!(is_cached(&smmu, 1, 0), "PASID 0 TLB should be warm before invalidation");
    assert!(is_cached(&smmu, 1, 1), "PASID 1 TLB should be warm before invalidation");

    // Invalidate only PASID 0's CD
    smmu.submit_command(cfgi_cd_cmd(1, 0)).unwrap();
    smmu.process_command_queue().unwrap();

    // PASID 0 entry is gone from TLB — but the page mapping still exists so
    // a full retranslation succeeds.  What we verify is that the TLB was cleared
    // by checking the invalidation counter increased.
    let (_, _, inv_before) = (0u64, 0u64, smmu.get_invalidation_count());

    // Re-warm to confirm the mapping is still valid
    assert!(is_cached(&smmu, 1, 0), "Page mapping for PASID 0 must still be reachable after CD invalidation");

    // PASID 1 must not have been touched
    assert!(is_cached(&smmu, 1, 1), "PASID 1 TLB must survive CMD_CFGI_CD targeting PASID 0");

    let _ = inv_before; // used to silence unused-variable lint
}

/// CMD_CFGI_CD must increase the invalidation counter.
#[test]
fn test_cfgi_cd_increments_invalidation_count() {
    let smmu = make_smmu();
    setup_stream_with_pasids(&smmu, 1, 1);

    let before = smmu.get_invalidation_count();
    smmu.submit_command(cfgi_cd_cmd(1, 0)).unwrap();
    smmu.process_command_queue().unwrap();
    let after = smmu.get_invalidation_count();

    assert!(after > before, "CMD_CFGI_CD must increment the invalidation counter");
}

/// CMD_CFGI_CD must NOT invalidate TLB entries belonging to a different stream.
#[test]
fn test_cfgi_cd_does_not_affect_other_streams() {
    let smmu = make_smmu();
    setup_stream_with_pasids(&smmu, 1, 1);
    setup_stream_with_pasids(&smmu, 2, 1);

    // Invalidate stream 1 / pasid 0
    smmu.submit_command(cfgi_cd_cmd(1, 0)).unwrap();
    smmu.process_command_queue().unwrap();

    // Stream 2's entries must survive
    assert!(is_cached(&smmu, 2, 0), "Stream 2's TLB entries must survive CMD_CFGI_CD targeting stream 1");
}

// ── CMD_CFGI_CD_ALL — all PASIDs for a stream ────────────────────────────────

/// CMD_CFGI_CD_ALL for stream 1 must invalidate TLB entries for all PASIDs of
/// that stream.
#[test]
fn test_cfgi_cd_all_invalidates_all_pasids() {
    let smmu = make_smmu();
    setup_stream_with_pasids(&smmu, 1, 3); // pasids 0, 1, 2

    let before = smmu.get_invalidation_count();

    smmu.submit_command(cfgi_cd_all_cmd(1)).unwrap();
    smmu.process_command_queue().unwrap();

    // Invalidation counter must increase
    assert!(smmu.get_invalidation_count() > before,
            "CMD_CFGI_CD_ALL must increment the invalidation counter");

    // All page mappings remain valid (the AddressSpace is not deleted, only
    // the TLB cache entries are evicted — re-translation must succeed).
    for p in 0..3u32 {
        assert!(is_cached(&smmu, 1, p),
                "PASID {} mapping must still be reachable after CMD_CFGI_CD_ALL (re-translation)", p);
    }
}

/// CMD_CFGI_CD_ALL must increase the invalidation counter.
#[test]
fn test_cfgi_cd_all_increments_invalidation_count() {
    let smmu = make_smmu();
    setup_stream_with_pasids(&smmu, 1, 2);

    let before = smmu.get_invalidation_count();
    smmu.submit_command(cfgi_cd_all_cmd(1)).unwrap();
    smmu.process_command_queue().unwrap();

    assert!(smmu.get_invalidation_count() > before,
            "CMD_CFGI_CD_ALL must increment the invalidation counter");
}

/// CMD_CFGI_CD_ALL for stream 1 must NOT invalidate TLB entries for stream 2.
#[test]
fn test_cfgi_cd_all_does_not_affect_other_streams() {
    let smmu = make_smmu();
    setup_stream_with_pasids(&smmu, 1, 1);
    setup_stream_with_pasids(&smmu, 2, 1);

    // Invalidate all of stream 1
    smmu.submit_command(cfgi_cd_all_cmd(1)).unwrap();
    smmu.process_command_queue().unwrap();

    // Stream 2 entries must survive
    assert!(is_cached(&smmu, 2, 0),
            "Stream 2's TLB entries must survive CMD_CFGI_CD_ALL targeting stream 1");
}

// ── CommandEntry::new() defaults ─────────────────────────────────────────────

/// CommandEntry::new() helper must produce valid CfgiCd and CfgiCdAll commands.
#[test]
fn test_command_entry_new_cfgi_cd() {
    let cmd = CommandEntry::new(CommandType::CfgiCd, 5, 3);
    assert_eq!(cmd.cmd_type, CommandType::CfgiCd);
    assert_eq!(cmd.stream_id, 5);
    assert_eq!(cmd.pasid, 3);
}

#[test]
fn test_command_entry_new_cfgi_cd_all() {
    let cmd = CommandEntry::new(CommandType::CfgiCdAll, 7, 0);
    assert_eq!(cmd.cmd_type, CommandType::CfgiCdAll);
    assert_eq!(cmd.stream_id, 7);
}
